src/jvm/main/org/jetbrains/kotlinx/lincheck/paramgen/ExpandingRangeIntGenerator.kt

```diff
@@ -0,0 +1,158 @@
+     */
+    private var currentStartInclusive: Int
+
+    /**
+     * Current right (inclusive) bound of current range
+     */
+    private var currentEndInclusive: Int
+
+    private var nextExpansionDirection = NextExpansionDirection.UP
+
+
+    init {
+        currentStartInclusive = startInclusive
+        currentEndInclusive = endInclusive
+    }
+
+    /**
+     * Reset bounds of this expanding range
+     */
+    fun restart() {
+        currentStartInclusive = startInclusive
+        currentEndInclusive = endInclusive
+    }
+
+    /**
+     * Generates next random number.
+     * With 50% percent probability expands current range if can and returns an expanded bound as a value.
+     * Otherwise, returns a random value from the current range
+     *
+     * @return random value from current range or moved bound
+     */
+    fun nextInt(): Int {
+        checkRangeExpansionAbility()
+        if (nextExpansionDirection == NextExpansionDirection.DISABLED || !random.nextBoolean()) {
+            return generateFromRandomRange(currentStartInclusive, currentEndInclusive)
+        }
+
+        return if (nextExpansionDirection == NextExpansionDirection.DOWN) {
+            nextExpansionDirection = NextExpansionDirection.UP
+            --currentStartInclusive
+        } else {
+            nextExpansionDirection = NextExpansionDirection.DOWN
+            ++currentEndInclusive
+        }
+    }
+
+    private fun checkRangeExpansionAbility() {
+        if (currentStartInclusive == minStartInclusive && currentEndInclusive == maxEndInclusive) {
+            nextExpansionDirection = NextExpansionDirection.DISABLED
+            return
+        }
+        if (currentStartInclusive == minStartInclusive) {
+            nextExpansionDirection = NextExpansionDirection.UP
+            return
+        }
+        if (currentEndInclusive == maxEndInclusive) {
+            nextExpansionDirection = NextExpansionDirection.DOWN
+        }
+    }
+
+    private enum class NextExpansionDirection {
+        UP,
+        DOWN,
+        DISABLED
+    }
+
+    private fun generateFromRandomRange(rangeLowerBoundInclusive: Int, rangeUpperBoundInclusive: Int): Int {
+        return rangeLowerBoundInclusive + random.nextInt(rangeUpperBoundInclusive - rangeLowerBoundInclusive + 1)
+    }
+
+    companion object {
+
+        private const val DEFAULT_START_VALUE = 0
+
+        /**
+         * Creates extensible range-based random generator.
+         *
+         * @param random        random for this generator
+         * @param configuration range configuration in format start:end, may be empty
+         * @param minStart      minimal range left bound (inclusive)
+         * @param maxEnd        maximal range right bound (inclusive)
+         * @param type          name of current type for generation for exception message if such appears
+         * @return created range-based random generator§
+         */
+        @JvmStatic
+        fun create(
```

Comment:
Yes, I'll create `ExpandingRangeIntGenerator`. I separated parsing logic of config parsing, because sometimes we need to call constructor explicitly with certain values, and it would look weird if we had to create a config string for it.